gcp/api/server.py

```diff
@@ -105,6 +108,66 @@ def QueryAffectedBatch(self, request, context):
 
     return osv_service_v1_pb2.BatchVulnerabilityList(results=batch_results)
 
+  @ndb_context
+  def DetermineVersion(self, request, context):
+    """Determine the version of the provided hashes."""
+    if not request.query.name:
+      context.abort(grpc.StatusCode.NOT_IMPLEMENTED,
+                    'Querying only by file hash is not implemented yet.')
+      return None
+    return get_version_by_name(request.query, context).result()
+
+
+@ndb.tasklet
+def get_version_by_name(version_query: osv_service_v1_pb2.VersionQuery,
+                        context: grpc.ServicerContext) -> ndb.Future:
+  """Identifies the version based on the provided name."""
+  query = osv.RepoIndex.query(osv.RepoIndex.name == version_query.name)
+  it = query.iter()
+  futures = []
+  while (yield it.has_next_async()):
+    idx = it.next()
+    match = compare_hashes_from_commit(idx, version_query.file_hashes)
+    futures.append(match)
+  results = []
+  for f in futures:
+    match = f.result()
+    if match.score != 0.0:
+      results.append(match)
+  return osv_service_v1_pb2.VersionMatchList(matches=results)
+
+
+@ndb.tasklet
+def compare_hashes_from_commit(
+    idx: osv.RepoIndex,
+    hashes: List[osv_service_v1_pb2.FileHash]) -> ndb.Future:
+  """"Retrieves the hashes from the provided index and compares them to the input hashes."""
+  total_files = 0
+  matching_hashes = 0
+  for i in range(idx.pages):
+    key = version_hashes_key(idx.key, idx.commit, idx.file_hash_type, i)
+    q = osv.RepoIndexResult.query(osv.RepoIndexResult.key == key)
+    it = q.iter()
+    while (yield it.has_next_async()):
+      res = it.next()
+      for idx_hash in res.file_results:
+        for in_hash in hashes:
+          if in_hash.hash == idx_hash:
```

Comment:
This would need to be changed to `idx_hash.hash` after the models.py change below.